README.md
NOTE: This project is now managed by the [OCS Organization](https://github.com/observatorycontrolsystem/). This repo is left as read-only.

You can access the new repo at [https://github.com/observatorycontrolsystem/thumbnail-service/](https://github.com/observatorycontrolsystem/thumbnail-service/)

# Thumbnail service

This is a flask application that generates thumbnails from data stored on the
[LCOGT Science Archive](https://developers.lcogt.net/pages/archive.html).

## Requirements

This script uses amazon S3 to temporarily store images. You'll need a bucket and the
[correct environmental variables](http://boto3.readthedocs.io/en/latest/guide/quickstart.html#configuration)
set.

This script also relies on [fits2image](https://github.com/LCOGT/fits2image). Make sure it is installed
in your virtualenv.

All other dependencies are listed in the pyproject.toml file and can be installed via [poetry](https://python-poetry.org/).

```bash
pip install --upgrade pip
pip install poetry
poetry install
```

`poetry install` will create a virtual environment for you, which will have all requirements installed.

Run the tests with `poetry run pytest thumbservice/tests.py`.

## Configuration

This project can be configured using the following environment variables:

|Environment Variable| Description | Default |
|--------------------|-------------|---------|
| `ARCHIVE_API_URL` | Archive API URL | http://localhost/
| `TMP_DIR` | Temporary directory for image operations | '/tmp/'
| `AWS_BUCKET` | AWS S3 Bucket to store thumbnails | 'changeme'
| `AWS_ACCESS_KEY_ID` | AWS Access Key ID for S3 Bucket | 'changeme'
| `AWS_SECRET_ACCESS_KEY` | AWS Secret Access Key for S3 Bucket | 'changeme'
| `STORAGE_URL` | URL for bucket storage if different than AWS. Defaults to None to connect to AWS S3 | None
| `REQUIRED_FRAME_VALIDATION_KEYS` | Keys from Archive API record required in order to create a thumbnail from the FITS image | 'configuration_type,request_id,filename'
| `VALID_CONFIGURATION_TYPES` | Only generate thumbnails from images of these configuration types | 'ARC,BIAS,BPM,DARK,DOUBLE,EXPERIMENTAL,EXPOSE,GUIDE,LAMPFLAT,SKYFLAT,SPECTRUM,STANDARD,TARGET,TRAILED'
| `VALID_CONFIGURATION_TYPES_FOR_COLOR_THUMBS` | Only generate color thumbnails from images of these configuration types | 'EXPOSE,STANDARD'

## Authorization

The API passes through the `Authorization` header to the archive API. You only need to provide
this header if the data you are attempting to get a thumbnail for is proprietary. See the archive
documentation for more details.

## Endpoints

There are only 2 endpoints: `/<frame_id>/` and `/<basename>/` where `frame_id` is the ID of the frame
in the archive, and `basename` is the base part of the filename (no file extension) you wish to make
a thumbnail of. Using the frame_id is faster to return if you happen to know it
ahead of time.

Both endpoints take the following query parameters:

* width
* height
* label
* image

Width and height are in pixels, label will appear as white text in the lower left had corner of the image.

They both **return a url** to the thumbnail file that will be good for 1 week unless the `image` parameter
is supplied which will return an image directly.


## Example

[https://thumbnails.lcogt.net/3863274/?width=500&height=500&label=So%20many%20stars](https://thumbnails.lcogt.net/3863274/?width=500&height=500&label=So%20many%20stars)

## Changelog

### 2.16
2020-02-15
* Fix hash used in s3 key to be stable given the same inputs
* Add hook to gunicorn configuration files to clear out the temp directory on startup

### 2.15
2020-01-03
* Clean up gunicorn config file as most configuration has been moved to the helm chart

### 2.14
2019-05-08
* Migrate to AWS V4 signatures
* Modernize docker image

### 2.13
2019-04-11
* Enable gunicorn error and access logs

### 2.12
2019-04-11
* Add `/robots.txt` and `/favicon.ico` endpoints
* Use gunicorn configuration file with `child_exit`
[server hook](http://docs.gunicorn.org/en/stable/settings.html#child-exit) implemented for additional
temp file cleanup actions on certain abrupt worker death [#846](https://lcoglobal.redmineup.com/issues/846)

### 2.11
2019-04-05
* Update boto3, requests, and flask
* Temporary file cleanup improvements
* Validate frame before attempting thumbnail generation
* Update error handling

### 2.10
2019-02-16
* Update numpy version
* Add DejaVu True Type Fonts
* Add freetype dependency for text overlay support
* Use standard LCO Jenkins automated docker container build pipeline
* Migrate Docker container to Alpine Linux base

### 2.9
2018-11-13
* Upgrade requests
* Initialize a variable that could have been referenced before assignment

### 2.5
* 2017-05-15
* Upgrade Dockerfile to be more sensible, upgrade to python 3.6

### 2.4
2017-04-04
* Add better exception handling to clean up temporary files even if generation fails.

### 2.3
2017-03-17
* Add quality paramter and reduce default quality. Improves speed without a noticeable change in image quality.
* Improve filename key so that any differnt combination of parameters will reuslt in a new image.

### 2.2
2017-02-27
* Added median filtering support

### 2.1
2016-08-24
* Fixed bug where images of different reduction levels were used to compose color images

### 2.0
2016-08-17
* Added color image support! Use ?color=True for frames which belong to a request that have other exposures
using red, visual and blue filters.

### 1.1
2016-08-11
* Added `image` url parameter to return image directly instead of json. Useful
for using as the src attribute in img tags.


CHANGELOG.md
# Changelog

This project adheres to semantic versioning.

## [Unreleased]
### Added
### Changed
### Removed


## [3.0.0] - 2022-04-11

Initial public release.

### Changed
- Update to use Flask v2
- Update to use Poetry for dependency management
- Generalize configuration to pull from environment variables with sensible defaults (see thumbservice/common.py)

## Pre-release changelog

### 2.16
2020-02-15
* Fix hash used in s3 key to be stable given the same inputs
* Add hook to gunicorn configuration files to clear out the temp directory on startup

### 2.15
2020-01-03
* Clean up gunicorn config file as most configuration has been moved to the helm chart

### 2.14
2019-05-08
* Migrate to AWS V4 signatures
* Modernize docker image

### 2.13
2019-04-11
* Enable gunicorn error and access logs

### 2.12
2019-04-11
* Add `/robots.txt` and `/favicon.ico` endpoints
* Use gunicorn configuration file with `child_exit`
[server hook](http://docs.gunicorn.org/en/stable/settings.html#child-exit) implemented for additional
temp file cleanup actions on certain abrupt worker death [#846](https://lcoglobal.redmineup.com/issues/846)

### 2.11
2019-04-05
* Update boto3, requests, and flask
* Temporary file cleanup improvements
* Validate frame before attempting thumbnail generation
* Update error handling

### 2.10
2019-02-16
* Update numpy version
* Add DejaVu True Type Fonts
* Add freetype dependency for text overlay support
* Use standard LCO Jenkins automated docker container build pipeline
* Migrate Docker container to Alpine Linux base

### 2.9
2018-11-13
* Upgrade requests
* Initialize a variable that could have been referenced before assignment

### 2.5
* 2017-05-15
* Upgrade Dockerfile to be more sensible, upgrade to python 3.6

### 2.4
2017-04-04
* Add better exception handling to clean up temporary files even if generation fails.

### 2.3
2017-03-17
* Add quality paramter and reduce default quality. Improves speed without a noticeable change in image quality.
* Improve filename key so that any differnt combination of parameters will reuslt in a new image.

### 2.2
2017-02-27
* Added median filtering support

### 2.1
2016-08-24
* Fixed bug where images of different reduction levels were used to compose color images

### 2.0
2016-08-17
* Added color image support! Use ?color=True for frames which belong to a request that have other exposures
using red, visual and blue filters.

### 1.1
2016-08-11
* Added `image` url parameter to return image directly instead of json. Useful
for using as the src attribute in img tags.


tests.py
import mock
from pathlib import Path
from copy import deepcopy
from collections import namedtuple

import boto3
import pytest
import requests
from moto import mock_s3

import common
import thumbservice

TEST_API_URL = 'https://test-archive-api.lco.gtn/'
TEST_BUCKET = 'test_bucket'
TEST_ACCESS_KEY = 'test_secret_key'
TEST_SECRET_ACCESS_KEY = 'test_secret_access_key'

_test_data = {
    'frame': {
        'OBSTYPE': 'EXPOSE',
        'filename': 'ogg0m404-kb82-20190321-0273-e91.fits.fz',
        'id': 11245132,
        'url': 'http://file_url',
        'PROPID': 'LCOEPO2018B-002',
        'REQNUM': 1756835,
        'FILTER': 'B'
    },
    'request_frames': {
        'count': 6,
        'results': [
            {
                'OBSTYPE': 'EXPOSE',
                'filename': 'ogg0m404-kb82-20190321-0273-e91.fits.fz',
                'id': 11245132,
                'url': 'http://file_url_1',
                'PROPID': 'LCOEPO2018B-002',
                'REQNUM': 1756835,
                'FILTER': 'B',
                'RLEVEL': 91
            },
            {
                'OBSTYPE': 'EXPOSE',
                'filename': 'ogg0m404-kb82-20190321-0273-e00.fits.fz',
                'id': 11245129,
                'url': 'http://file_url_2',
                'PROPID': 'LCOEPO2018B-002',
                'REQNUM': 1756835,
                'FILTER': 'B',
                'RLEVEL': 0
            },
            {
                'OBSTYPE': 'EXPOSE',
                'filename': 'ogg0m404-kb82-20190321-0272-e91.fits.fz',
                'id': 11245120,
                'url': 'http://file_url_3',
                'PROPID': 'LCOEPO2018B-002',
                'REQNUM': 1756835,
                'FILTER': 'V',
                'RLEVEL': 91
            },
            {
                'OBSTYPE': 'EXPOSE',
                'filename': 'ogg0m404-kb82-20190321-0272-e00.fits.fz',
                'id': 11245119,
                'url': 'http://file_url_4',
                'PROPID': 'LCOEPO2018B-002',
                'REQNUM': 1756835,
                'FILTER': 'V',
                'RLEVEL': 0
            },
            {
                'OBSTYPE': 'EXPOSE',
                'filename': 'ogg0m404-kb82-20190321-0271-e91.fits.fz',
                'id': 11245105,
                'url': 'http://file_url_5',
                'PROPID': 'LCOEPO2018B-002',
                'REQNUM': 1756835,
                'FILTER': 'rp',
                'RLEVEL': 91
            },
            {
                'OBSTYPE': 'EXPOSE',
                'filename': 'ogg0m404-kb82-20190321-0271-e00.fits.fz',
                'id': 11245103,
                'url': 'http://file_url_6',
                'PROPID': 'LCOEPO2018B-002',
                'REQNUM': 1756835,
                'FILTER': 'rp',
                'RLEVEL': 0
            },
        ]
    }
}


@pytest.fixture(autouse=True)
def set_test_values(tmp_path):
    thumbservice.settings = common.Settings(
        settings={
            'TMP_DIR': tmp_path,
            'ARCHIVE_API': TEST_API_URL,
            'AWS_S3_BUCKET': TEST_BUCKET,
            'AWS_ACCESS_KEY_ID': TEST_ACCESS_KEY,
            'AWS_SECRET_ACCESS_KEY': TEST_SECRET_ACCESS_KEY
        }
    )


@pytest.fixture(autouse=True)
def mock_fits_to_jpeg():
    def side_effect(*args, **kwargs):
        Path(args[1]).touch()
    m = thumbservice.fits_to_jpg = mock.MagicMock()
    m.side_effect = side_effect


def make_tmp_file(tmp_path, filename, suffix):
    path = tmp_path / Path(filename).with_suffix(suffix)
    path.touch()
    return str(path)


@pytest.fixture(autouse=True)
def mock_affineremap(tmp_path, request):
    if 'no_auto_mock_affineremap' in request.keywords:
        return

    def side_effect(*args, **kwargs):
        return make_tmp_file(tmp_path, args[0], '.affineremap')

    m = thumbservice.affineremap = mock.MagicMock()
    m.side_effect = side_effect


def make_transforms_returns(paths, is_ok):
    ukn = namedtuple('Ukn', ['filepath'])
    result = namedtuple('Result', ['ok', 'trans', 'ukn'])
    results = []
    for path in paths:
        results.append(result(is_ok, None, ukn(path)))
    return results


@pytest.fixture(autouse=True)
def mock_make_transforms(request):
    if 'no_auto_mock_make_transforms' in request.keywords:
        return

    def side_effect(*args, **kwargs):
        return make_transforms_returns(args[1], is_ok=True)

    m = thumbservice.make_transforms = mock.MagicMock()
    m.side_effect = side_effect


@pytest.fixture
def thumbservice_client():
    thumbservice.app.config['TESTING'] = True
    yield thumbservice.app.test_client()


@pytest.fixture
def s3_client():
    # This should be passed in to all test functions to mock out calls to aws
    with mock_s3():
        config = boto3.session.Config(region_name='us-west-2', signature_version='s3v4')
        s3 = boto3.client('s3', aws_access_key_id=TEST_ACCESS_KEY, aws_secret_access_key=TEST_SECRET_ACCESS_KEY, config=config)
        s3.create_bucket(Bucket=TEST_BUCKET)
        yield s3


def test_get_index(thumbservice_client):
    response = thumbservice_client.get('/')
    assert b'Please see the documentation' in response.data


def test_get_index_with_random_path(thumbservice_client):
    response = thumbservice_client.get('/some/random/endpoint')
    assert b'Please see the documentation' in response.data


def test_generate_black_and_white_thumbnail_successfully(thumbservice_client, requests_mock, s3_client, tmp_path):
    frame = deepcopy(_test_data['frame'])
    requests_mock.get(f'{TEST_API_URL}frames/{frame["id"]}/', json=frame)
    requests_mock.get(frame['url'], content=b'I Am Image')
    response1 = thumbservice_client.get(f'/{frame["id"]}/')
    call_count_after_1 = requests_mock.call_count
    response2 = thumbservice_client.get(f'/{frame["id"]}/')
    call_count_after_2 = requests_mock.call_count
    for response in [response1, response2]:
        response_as_json = response.get_json()
        assert response_as_json['propid'] == frame['PROPID']
        assert 'url' in response_as_json
        assert response.status_code == 200
    # The resource will have been created in s3 on the first call, on the second call less work needs to
    # be done, including only 1 call to requests as opposed to the 2 on initial creation
    assert call_count_after_1 == 2
    assert call_count_after_2 == 3
    # All temp files should have been cleared out
    assert len(list(tmp_path.glob('*'))) == 0


def test_generate_color_thumbnail_successfully(thumbservice_client, requests_mock, s3_client, tmp_path):
    frame = deepcopy(_test_data['frame'])
    request_frames = deepcopy(_test_data['request_frames'])
    requests_mock.get(f'{TEST_API_URL}frames/{frame["id"]}/', json=frame)
    requests_mock.get(f'{TEST_API_URL}frames/?REQNUM={frame["REQNUM"]}&RLEVEL=91', json=request_frames)
    for request_frame in request_frames['results']:
        requests_mock.get(request_frame['url'], content=b'I Am Image')
    response1 = thumbservice_client.get(f'/{frame["id"]}/?color=true')
    call_count_after_1 = requests_mock.call_count
    response2 = thumbservice_client.get(f'/{frame["id"]}/?color=true')
    call_count_after_2 = requests_mock.call_count
    for response in [response1, response2]:
        response_as_json = response.get_json()
        assert response_as_json['propid'] == frame['PROPID']
        assert 'url' in response_as_json
        assert response.status_code == 200
    # The resource will have been created in s3 on the first call, on the second call less work needs to
    # be done, including only 1 call to requests as opposed to the 5 on initial creation
    assert call_count_after_1 == 5
    assert call_count_after_2 == 6
    # All temp files should have been cleared out
    assert len(list(tmp_path.glob('*'))) == 0


@pytest.mark.no_auto_mock_affineremap
def test_image_align_fails_falls_back_to_original_image_list(thumbservice_client, requests_mock, tmp_path, s3_client):
    m = thumbservice.affineremap = mock.MagicMock()
    m.side_effect = [make_tmp_file(tmp_path, 'tmp', '.fits'), Exception('Something bad happened')]
    frame = deepcopy(_test_data['frame'])
    request_frames = deepcopy(_test_data['request_frames'])
    requests_mock.get(f'{TEST_API_URL}frames/{frame["id"]}/', json=frame)
    requests_mock.get(f'{TEST_API_URL}frames/?REQNUM={frame["REQNUM"]}&RLEVEL=91', json=request_frames)
    for request_frame in request_frames['results']:
        requests_mock.get(request_frame['url'], content=b'I Am Image')
    response = thumbservice_client.get(f'/{frame["id"]}/?color=true')
    response_as_json = response.get_json()
    assert 'url' in response_as_json
    assert response_as_json['propid'] == frame['PROPID']
    assert response.status_code == 200
    assert len(list(tmp_path.glob('*'))) == 0


@pytest.mark.mock_make_transforms
def test_one_image_doesnt_align_falls_back_to_original_image_list(thumbservice_client, requests_mock, tmp_path, s3_client):
    m = thumbservice.make_transforms = mock.MagicMock()
    m.side_effect = [make_transforms_returns(['a.fits'], True), make_transforms_returns(['b.fits'], False)]
    frame = deepcopy(_test_data['frame'])
    request_frames = deepcopy(_test_data['request_frames'])
    requests_mock.get(f'{TEST_API_URL}frames/{frame["id"]}/', json=frame)
    requests_mock.get(f'{TEST_API_URL}frames/?REQNUM={frame["REQNUM"]}&RLEVEL=91', json=request_frames)
    for request_frame in request_frames['results']:
        requests_mock.get(request_frame['url'], content=b'I Am Image')
    response = thumbservice_client.get(f'/{frame["id"]}/?color=true')
    response_as_json = response.get_json()
    assert 'url' in response_as_json
    assert response_as_json['propid'] == frame['PROPID']
    assert response.status_code == 200
    assert len(list(tmp_path.glob('*'))) == 0


def test_all_filters_for_color_thumbnail_not_available(thumbservice_client, requests_mock, s3_client, tmp_path):
    frame = deepcopy(_test_data['frame'])
    request_frames = deepcopy(_test_data['request_frames'])
    request_frames['results'].pop()
    request_frames['results'].pop()
    requests_mock.get(f'{TEST_API_URL}frames/{frame["id"]}/', json=frame)
    requests_mock.get(f'{TEST_API_URL}frames/?REQNUM={frame["REQNUM"]}&RLEVEL=91', json=request_frames)
    response = thumbservice_client.get(f'/{frame["id"]}/?color=true')
    assert response.status_code == 404
    assert b'RVB frames not found' in response.data
    assert len(list(tmp_path.glob('*'))) == 0


def test_reduced_frames_for_color_thumbnail_not_available(thumbservice_client, requests_mock, s3_client, tmp_path):
    frame = deepcopy(_test_data['frame'])
    requests_mock.get(f'{TEST_API_URL}frames/{frame["id"]}/', json=frame)
    requests_mock.get(f'{TEST_API_URL}frames/?REQNUM={frame["REQNUM"]}&RLEVEL=91', json={'results': []})
    response = thumbservice_client.get(f'/{frame["id"]}/?color=true')
    assert response.status_code == 404
    assert b'RVB frames not found' in response.data
    assert len(list(tmp_path.glob('*'))) == 0


def test_cannot_generate_thumbnail_for_non_image_obstypes(thumbservice_client, requests_mock, tmp_path, s3_client):
    frame = deepcopy(_test_data['frame'])
    frame['OBSTYPE'] = 'CATALOG'
    requests_mock.get(f'{TEST_API_URL}frames/{frame["id"]}/', json=frame)
    response = thumbservice_client.get(f'/{frame["id"]}/')
    assert response.status_code == 400
    assert 'Cannot generate thumbnail for OBSTYPE=CATALOG' in response.get_json()['message']
    assert len(list(tmp_path.glob('*'))) == 0


def test_cannot_generate_color_thumbnail_for_all_valid_obstypes(thumbservice_client, requests_mock, tmp_path, s3_client):
    frame = deepcopy(_test_data['frame'])
    frame['OBSTYPE'] = 'SPECTRUM'
    requests_mock.get(f'{TEST_API_URL}frames/{frame["id"]}/', json=frame)
    response = thumbservice_client.get(f'/{frame["id"]}/?color=true')
    assert response.status_code == 400
    assert 'Cannot generate color thumbnail for OBSTYPE=SPECTRUM' in response.get_json()['message']
    assert len(list(tmp_path.glob('*'))) == 0


def test_cannot_generate_thumbnail_for_non_fits_file(thumbservice_client, requests_mock, tmp_path, s3_client):
    frame = deepcopy(_test_data['frame'])
    frame['filename'] = 'OGG_calib_0001760408_ftn_20190331_58574.tar.gz'
    frame['OBSTYPE'] = 'SPECTRUM'
    requests_mock.get(f'{TEST_API_URL}frames/{frame["id"]}/', json=frame)
    response = thumbservice_client.get(f'/{frame["id"]}/')
    assert response.status_code == 400
    assert 'Cannot generate thumbnail for non FITS-type frame' in response.get_json()['message']
    assert len(list(tmp_path.glob('*'))) == 0


def test_cannot_generate_color_thumbnail_not_associated_with_a_request(thumbservice_client, requests_mock, tmp_path, s3_client):
    frame = deepcopy(_test_data['frame'])
    frame['REQNUM'] = None
    requests_mock.get(f'{TEST_API_URL}frames/{frame["id"]}/', json=frame)
    response = thumbservice_client.get(f'/{frame["id"]}/?color=true')
    assert response.status_code == 400
    assert 'Cannot generate color thumbnail for a frame that does not have a request' in response.get_json()['message']
    assert len(list(tmp_path.glob('*'))) == 0


def test_cannot_generate_color_thumbnail_with_incomplete_frame_info(thumbservice_client, requests_mock, tmp_path, s3_client):
    frame = deepcopy(_test_data['frame'])
    del frame['REQNUM']
    requests_mock.get(f'{TEST_API_URL}frames/{frame["id"]}/', json=frame)
    response = thumbservice_client.get(f'/{frame["id"]}', follow_redirects=True)
    assert response.status_code == 400
    assert 'Cannot generate thumbnail for given frame' in response.get_json()['message']
    assert len(list(tmp_path.glob('*'))) == 0


def test_frame_not_found(thumbservice_client, requests_mock, tmp_path, s3_client):
    frame_id = 6
    frame = {
        'detail': 'Not found.'
    }
    requests_mock.get(f'{TEST_API_URL}frames/{frame_id}/', json=frame, status_code=404)
    response = thumbservice_client.get(f'/{frame_id}/')
    assert response.status_code == 404
    assert len(list(tmp_path.glob('*'))) == 0


def test_archive_query_returned_500(thumbservice_client, requests_mock, tmp_path, s3_client):
    frame_id = 13
    requests_mock.get(f'{TEST_API_URL}frames/{frame_id}/', status_code=500)
    response = thumbservice_client.get(f'/{frame_id}/')
    assert response.status_code == 502
    assert len(list(tmp_path.glob('*'))) == 0


def test_archive_query_raised_exception_during_request(thumbservice_client, requests_mock, tmp_path, s3_client):
    frame_id = 13
    requests_mock.get(f'{TEST_API_URL}frames/{frame_id}/', exc=requests.exceptions.ConnectTimeout)
    response = thumbservice_client.get(f'/{frame_id}/')
    assert response.status_code == 502
    assert len(list(tmp_path.glob('*'))) == 0


def test_frame_basename_does_not_exist(thumbservice_client, requests_mock, tmp_path, s3_client):
    empty_results = {
        'count': 0,
        'results': []
    }
    requests_mock.get(f'{TEST_API_URL}frames/?basename=some_frame_that_doesnt_exist', json=empty_results)
    response = thumbservice_client.get('/some_frame_that_doesnt_exist/')
    assert response.status_code == 404
    assert len(list(tmp_path.glob('*'))) == 0


thumbservice/common.py
import os


def get_temp_filename_prefix(pid=None):
    # Common prefix for all files created by a single process. Used to both create
    # files and find files from a given pid to clean up if necessary.
    pid = os.getpid() if pid is None else pid
    return f'pid{pid}-'


class Settings:
    def __init__(self, settings=None):
        self._settings = settings or {}
        self.ARCHIVE_API_URL = self.set_value('ARCHIVE_API_URL', 'http://localhost/', True)
        self.TMP_DIR = self.set_value('TMP_DIR', '/tmp/', True)
        self.AWS_BUCKET = self.set_value('AWS_BUCKET', 'changeme')
        self.AWS_ACCESS_KEY_ID = self.set_value('AWS_ACCESS_KEY_ID', 'changeme')
        self.AWS_SECRET_ACCESS_KEY = self.set_value('AWS_SECRET_ACCESS_KEY', 'changeme')
        self.STORAGE_URL = self.set_value('STORAGE_URL', None)
        self.REQUIRED_FRAME_VALIDATION_KEYS = self.get_tuple_from_environment('REQUIRED_FRAME_VALIDATION_KEYS', 'configuration_type,request_id,filename')
        self.VALID_CONFIGURATION_TYPES = self.get_tuple_from_environment('VALID_CONFIGURATION_TYPES', 'ARC,BIAS,BPM,DARK,DOUBLE,EXPERIMENTAL,EXPOSE,GUIDE,LAMPFLAT,SKYFLAT,SPECTRUM,STANDARD,TARGET,TRAILED')
        self.VALID_CONFIGURATION_TYPES_FOR_COLOR_THUMBS = self.get_tuple_from_environment('VALID_CONFIGURATION_TYPES_FOR_COLOR_THUMBS', 'EXPOSE,STANDARD')

    def set_value(self, env_var, default, must_end_with_slash=False):
        if env_var in self._settings:
            value = self._settings[env_var]
        else:
            value = os.getenv(env_var, default)
        return self.end_with_slash(value) if must_end_with_slash else value

    @staticmethod
    def end_with_slash(path):
        return os.path.join(path, '')

    def get_tuple_from_environment(self, variable_name, default):
        return tuple(os.getenv(variable_name, default).strip(',').replace(' ', '').split(','))

settings = Settings()



thumbservice/thumbservice.py
#!/usr/bin/env python
import os
import uuid
import logging
import hashlib

import boto3
import requests
from flask_cors import CORS
from flask.logging import default_handler
from flask import Flask, request, jsonify, redirect, send_from_directory
from fits2image.conversions import fits_to_jpg
from fits_align.ident import make_transforms
from fits_align.align import affineremap

from thumbservice.common import settings, get_temp_filename_prefix


app = Flask(__name__, static_folder='static')
CORS(app)

class RequestFormatter(logging.Formatter):
    def format(self, record):
        record.url = request.url
        return super().format(record)

formatter = RequestFormatter('[%(asctime)s] %(levelname)s in %(module)s for %(url)s: %(message)s')
default_handler.setFormatter(formatter)


class ThumbnailAppException(Exception):
    status_code = 500

    def __init__(self, message, status_code=None, payload=None):
        Exception.__init__(self)
        self.message = message
        if status_code is not None:
            self.status_code = status_code
        self.payload = payload

    def to_dict(self):
        result = dict(self.payload or ())
        result['message'] = self.message
        return result


@app.errorhandler(ThumbnailAppException)
def handle_thumbnail_app_exception(error):
    response = jsonify(error.to_dict())
    response.status_code = error.status_code
    return response


def get_response(url, params=None, headers=None):
    response = None
    try:
        response = requests.get(url, headers=headers, params=params, timeout=10)
        response.raise_for_status()
    except requests.RequestException:
        status_code = getattr(response, 'status_code', None)
        payload = {}
        message = 'Got error response'
        if status_code is None or 500 <= status_code < 600:
            status_code = 502
        elif status_code == 404:
            message = 'Not found'
        else:
            try:
                payload['response'] = response.json()
            except:
                pass
        raise ThumbnailAppException(message, status_code=status_code, payload=payload)
    return response


def can_generate_thumbnail_on(frame, request):
    frame_has_required_validation_keys = all([key in frame.keys() for key in settings.REQUIRED_FRAME_VALIDATION_KEYS])
    if not frame_has_required_validation_keys:
        return {'result': False, 'reason': 'Cannot generate thumbnail for given frame'}

    configuration_type = frame.get('configuration_type').upper()
    request_id = frame.get('request_id')
    is_color_request = request.args.get('color', 'false') == 'true'
    is_fits_file = any([frame.get('filename').endswith(ext) for ext in ['.fits', '.fits.fz']])

    if configuration_type not in settings.VALID_CONFIGURATION_TYPES:
        return {'result': False, 'reason': f'Cannot generate thumbnail for configuration_type={configuration_type}'}

    if is_color_request and not request_id:
        return {'result': False, 'reason': 'Cannot generate color thumbnail for a frame that does not have a request'}

    if is_color_request and configuration_type not in settings.VALID_CONFIGURATION_TYPES_FOR_COLOR_THUMBS:
        return {'result': False, 'reason': f'Cannot generate color thumbnail for configuration_type={configuration_type}'}

    if not is_fits_file:
        return {'result': False, 'reason': 'Cannot generate thumbnail for non FITS-type frame'}

    return {'result': True, 'reason': ''}


def unique_temp_path_start():
    return f'{settings.TMP_DIR}{get_temp_filename_prefix()}{uuid.uuid4().hex}-'


def save_temp_file(frame):
    path = f'{unique_temp_path_start()}{frame["filename"]}'
    with open(path, 'wb') as f:
        f.write(get_response(frame['url']).content)
    return path


def key_for_jpeg(frame_id, **params):
    return f'{frame_id}.{hashlib.blake2b(repr(frozenset(params.items())).encode(), digest_size=20).hexdigest()}.jpg'


def convert_to_jpg(paths, key, **params):
    jpg_path = f'{unique_temp_path_start()}{key}'
    fits_to_jpg(paths, jpg_path, **params)
    return jpg_path


def get_s3_client():
    config = boto3.session.Config(region_name='us-west-2', signature_version='s3v4')
    return boto3.client(
        's3',
        aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
        aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
        endpoint_url=settings.STORAGE_URL,
        config=config,
    )


def upload_to_s3(key, jpg_path):
    client = get_s3_client()
    with open(jpg_path, 'rb') as f:
        client.put_object(
            Bucket=settings.AWS_BUCKET,
            Body=f,
            Key=key,
            ContentType='image/jpeg'
        )


def generate_url(key):
    client = get_s3_client()
    return client.generate_presigned_url(
        'get_object',
        ExpiresIn=3600 * 8,
        Params={'Bucket': settings.AWS_BUCKET, 'Key': key}
    )


def key_exists(key):
    client = get_s3_client()
    try:
        client.head_object(Bucket=settings.AWS_BUCKET, Key=key)
        return True
    except:
        return False


def frames_for_requestnum(request_id, request, reduction_level):
    headers = {
        'Authorization': request.headers.get('Authorization')
    }
    params = {'request_id': request_id, 'reduction_level': reduction_level}
    return get_response(f'{settings.ARCHIVE_API_URL}frames/', params=params, headers=headers).json()['results']


def rvb_frames(frames):
    FILTERS_FOR_COLORS = {
        'red': ['R', 'rp'],
        'visual': ['V'],
        'blue': ['B'],
    }
    selected_frames = []
    for color in ['red', 'visual', 'blue']:
        try:
            selected_frames.append(
                next(f for f in frames if f['primary_optical_element'] in FILTERS_FOR_COLORS[color])
            )
        except StopIteration:
            raise ThumbnailAppException('RVB frames not found', status_code=404)
    return selected_frames


def reproject_files(ref_image, images_to_align):
    """Return three aligned images."""
    aligned_images = []
    reprojected_file_list = [ref_image]
    try:
        identifications = make_transforms(ref_image, images_to_align[1:3])
        for id in identifications:
            if id.ok:
                aligned_image = affineremap(id.ukn.filepath, id.trans, outdir=settings.TMP_DIR)
                aligned_images.append(aligned_image)
    except Exception:
        app.logger.warning('Error aligning images, falling back to original image list', exc_info=True)

    # Clean up aligned images if they will not be used
    if len(aligned_images) != 2:
        while len(aligned_images) > 0:
            aligned_image = aligned_images.pop()
            if os.path.exists(aligned_image):
                os.remove(aligned_image)

    reprojected_file_list = reprojected_file_list + aligned_images
    return reprojected_file_list if len(reprojected_file_list) == 3 else images_to_align


class Paths:
    """Retain all paths set"""
    def __init__(self):
        self._all_paths = set()
        self.paths = []

    def set(self, paths):
        for path in paths:
            self._all_paths.add(path)
        self.paths = paths

    @property
    def all_paths(self):
        return list(self._all_paths)


def generate_thumbnail(frame, request):
    params = {
        'width': int(request.args.get('width', 200)),
        'height': int(request.args.get('height', 200)),
        'label_text': request.args.get('label'),
        'color': request.args.get('color', 'false') != 'false',
        'median': request.args.get('median', 'false') != 'false',
        'percentile': float(request.args.get('percentile', 99.5)),
        'quality': int(request.args.get('quality', 80)),
    }
    key = key_for_jpeg(frame['id'], **params)
    if key_exists(key):
        return generate_url(key)
    # Cfitsio is a bit crappy and can only read data off disk
    jpg_path = None
    paths = Paths()
    try:
        if not params['color']:
            paths.set([save_temp_file(frame)])
        else:
            # Color thumbnails can only be generated on rlevel 91 images
            reqnum_frames = frames_for_requestnum(frame['request_id'], request, reduction_level=91)
            paths.set([save_temp_file(frame) for frame in rvb_frames(reqnum_frames)])
            paths.set(reproject_files(paths.paths[0], paths.paths))
        jpg_path = convert_to_jpg(paths.paths, key, **params)
        upload_to_s3(key, jpg_path)
    finally:
        # Cleanup actions
        if jpg_path and os.path.exists(jpg_path):
            os.remove(jpg_path)
        for path in paths.all_paths:
            if os.path.exists(path):
                os.remove(path)
    return generate_url(key)


def handle_response(frame, request):
    can_generate_thumbnail_on_frame = can_generate_thumbnail_on(frame, request)
    if not can_generate_thumbnail_on_frame['result']:
        raise ThumbnailAppException(can_generate_thumbnail_on_frame['reason'], status_code=400)

    url = generate_thumbnail(frame, request)
    if request.args.get('image'):
        return redirect(url)
    else:
        return jsonify({'url': url, 'propid': frame['proposal_id']})


@app.route('/<frame_basename>/')
def bn_thumbnail(frame_basename):
    headers = {
        'Authorization': request.headers.get('Authorization')
    }
    params = {'basename': frame_basename}
    frames = get_response(f'{settings.ARCHIVE_API_URL}frames/', params=params, headers=headers).json()

    if not frames['count'] == 1:
        raise ThumbnailAppException('Not found', status_code=404)

    return handle_response(frames['results'][0], request)


@app.route('/<int:frame_id>/')
def thumbnail(frame_id):
    headers = {
        'Authorization': request.headers.get('Authorization')
    }
    frame = get_response(f'{settings.ARCHIVE_API_URL}frames/{frame_id}/', headers=headers).json()

    return handle_response(frame, request)


@app.route('/favicon.ico')
def favicon():
    return redirect('https://cdn.lco.global/mainstyle/img/favicon.ico')


@app.route('/robots.txt')
def robots():
    return send_from_directory(app.static_folder, 'robots.txt')


@app.route('/', defaults={'path': ''})
@app.route('/<path:path>')
def index(path):
    return ((
        'Please see the documentation for the thumbnail service at '
        '<a href="https://developers.lco.global">developers.lco.global</a>'
    ))

if __name__ == '__main__':
    app.run(host='0.0.0.0', debug=True)
